mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -34,33 +35,142 @@
+                            moduleOp, rewriter.getStringAttr(foldedName));
+                        if (funcOp->hasAttr("qnode") and !foldedOp) {
+                            // Create the folded circuit function
+                            auto foldedCircuitAttr =
+                                getOrInsertFoldedCircuit(loc, rewriter, funcOp, foldingAlgorithm);
+                            func::FuncOp foldedOp =
+                                SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                                    moduleOp, foldedCircuitAttr);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.setInsertionPoint(callOp);
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
+                        }
+                        else if (funcOp->hasAttr("qnode") and foldedOp) {
+                            rewriter.setInsertionPoint(callOp);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
+                        }
+                    });
+                }
+            }
+        });
+
+        std::string fnName = calleeOp.getName().str() + ".zne";
+        FlatSymbolRefAttr foldedOpRefAttr = SymbolRefAttr::get(op.getContext(), fnName);
+        fnFoldedOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(moduleOp, foldedOpRefAttr);
+        // Traverse the call graph a second time, in orderd to replace the function calls to their
+        // .zne counterparts.
+        traverseCallGraph(fnFoldedOp, /*symbolTable=*/nullptr, [&](func::FuncOp funcOp) {
+            funcOp.walk([&](func::CallOp callOp) {
+                PatternRewriter::InsertionGuard insertionGuard(rewriter);
+                std::string fnName = callOp.getCallee().str() + ".zne";
+                FlatSymbolRefAttr foldedOpRefAttr = SymbolRefAttr::get(op.getContext(), fnName);
+                auto currentFnFoldedOp =
+                    SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(moduleOp, foldedOpRefAttr);
+                if (currentFnFoldedOp) {
+                    rewriter.setInsertionPointToStart(&currentFnFoldedOp.getFunctionBody().front());
+
+                    callOp.setCallee(currentFnFoldedOp.getName());
+                    auto parentFunc = callOp->getParentOfType<func::FuncOp>();
+                    callOp.getOperandsMutable().append(parentFunc.getArguments().back());
```

Comment:
I don't think you need to modify the IP in order to mutate an existing operation. But you do need to notify the rewriter that you are doing so.

Reply:
Updated, with modifyOpInPlace